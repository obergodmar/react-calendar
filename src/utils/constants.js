export const TEST_DATE = new Date(1970, 1, 1);
export const FULL_FORMAT = "MMddyyyy";
export const WEEKDAY_FORMAT = "EEEEEE";
export const MONTH_FORMAT = "MMM";
export const MONTH_YEAR = "LLLL yyyy";
export const DATE_FORMAT = "MM / dd / yyyy";
