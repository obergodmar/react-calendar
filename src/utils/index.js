export { default as dayMatrix } from "./day-matrix";
export * from "./constants";
