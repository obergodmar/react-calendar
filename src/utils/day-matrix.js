import { addDays, format, startOfWeek } from "date-fns";
import { FULL_FORMAT } from "./constants";

const rows = 6;
const cols = 7;
const length = rows * cols;

export default (currentDate) => {
  const today = format(new Date(), FULL_FORMAT);
  const currentDateFormat = format(currentDate, FULL_FORMAT);
  const year = currentDate.getFullYear();
  const month = currentDate.getMonth();
  const startDate = startOfWeek(new Date(year, month, 1));
  return Array.from({ length })
    .map((_, index) => {
      const date = addDays(startDate, index);
      const dateFormat = format(date, FULL_FORMAT);
      const currentMonth = date.getMonth();
      const isPart = currentMonth === month;
      const isToday = dateFormat === today;
      const isSelected = currentDateFormat === dateFormat;
      return {
        date,
        isToday,
        isSelected,
        isPart,
      };
    })
    .reduce(
      (matrix, _, index, days) =>
        !(index % cols !== 0)
          ? [...matrix, days.slice(index, index + cols)]
          : matrix,
      []
    );
};
