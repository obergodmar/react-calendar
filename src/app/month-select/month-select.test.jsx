import React from "react";
import { shallow } from "enzyme";
import { describe, jest, test } from "@jest/globals";

import { MonthSelect } from ".";
import { TEST_DATE } from "../../utils";

describe("Month Select", () => {
  const handleChangeMonth = jest.fn();
  const date = new Date(TEST_DATE);
  const wrapper = shallow(
    <MonthSelect date={date} changeMonth={handleChangeMonth} />
  );

  test("Should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Should change month by click", () => {
    wrapper.find("li").at(5).simulate("click");
    expect(handleChangeMonth).toHaveBeenCalledWith(6);
  });

  test("Should change month by keydown", () => {
    wrapper.find("li").at(5).simulate("keydown", { keyCode: 13 });
    expect(handleChangeMonth).toHaveBeenCalledWith(6);

    wrapper.find("li").at(7).simulate("keydown", { keyCode: 32 });
    expect(handleChangeMonth).toHaveBeenCalledWith(8);
  });

  test("Should not change month by keydown", () => {
    const newChangeMonth = jest.fn();
    const component = shallow(
      <MonthSelect date={date} changeMonth={newChangeMonth} />
    );

    component.find("li").at(5).simulate("keydown", { keyCode: 37 });
    expect(newChangeMonth).not.toHaveBeenCalled();
  });

  test("Should be default prevented", () => {
    const event = { preventDefault: jest.fn() };

    wrapper.find("li").at(5).simulate("mousedown", event);
    expect(event.preventDefault).toHaveBeenCalledTimes(1);
  });

  test("Should be 1st of January", () => {
    expect(wrapper.find("li").at(0).is(".monthSelectItemSelected")).toEqual(
      true
    );
  });

  test("Should be today's year", () => {
    const month = new Date().getMonth();
    const dateToday = new Date();
    const component = shallow(
      <MonthSelect date={dateToday} changeMonth={handleChangeMonth} />
    );

    expect(
      component
        .find("li")
        .at(month - 1)
        .is(".monthSelectItemToday")
    ).toEqual(true);
  });
});
