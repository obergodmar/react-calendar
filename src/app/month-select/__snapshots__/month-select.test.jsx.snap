// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Month Select Should match snapshot 1`] = `
<ul
  className="monthSelect"
>
  <li
    className="monthSelectItem monthSelectItemSelected"
    key="0"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Feb
  </li>
  <li
    className="monthSelectItem"
    key="1"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Mar
  </li>
  <li
    className="monthSelectItem"
    key="2"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Apr
  </li>
  <li
    className="monthSelectItem"
    key="3"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    May
  </li>
  <li
    className="monthSelectItem"
    key="4"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Jun
  </li>
  <li
    className="monthSelectItem"
    key="5"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Jul
  </li>
  <li
    className="monthSelectItem"
    key="6"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Aug
  </li>
  <li
    className="monthSelectItem"
    key="7"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Sep
  </li>
  <li
    className="monthSelectItem"
    key="8"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Oct
  </li>
  <li
    className="monthSelectItem"
    key="9"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Nov
  </li>
  <li
    className="monthSelectItem"
    key="10"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Dec
  </li>
  <li
    className="monthSelectItem"
    key="11"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    tabIndex={0}
  >
    Jan
  </li>
</ul>
`;
